src/Functions/FunctionsCharsetClassification.cpp

```diff
@@ -0,0 +1,247 @@
+
+        return len;
+    }
+
+
+    static void constant(String data, String & res)
+    {
+        const auto & encodings_freq = FrequencyHolder::getInstance().getEncodingsFrequency();
+
+        std::unordered_map<UInt16, Float64> model;
+        calculateStats(data.data(), data.size(), readCodePoints, model);
+
+        Float64 max_result = log(zero_frequency) * (max_string_size);
+        String poss_ans;
+        /// Go through the dictionary and find the charset with the highest weight
+        for (auto& item : encodings_freq)
+        {
+            Float64 score = naiveBayes(item.map, model, max_result);
+            if (max_result < score)
+            {
+                poss_ans = item.name;
+                max_result = score;
+            }
+        }
+
+        /* In our dictionary we have lines with form: <Language>_<Charset>
+         * If we need to find language of data, we return <Language>
+         * If we need to find charset of data, we return <Charset>.
+         */
+
+        size_t sep = poss_ans.find('_');
+        if (detect_language)
+        {
+            res = poss_ans.erase(0, sep + 1);
+        }
+        else
+        {
+            res = poss_ans.erase(sep, poss_ans.size() - sep);
+        }
+    }
+
+
+    static void vector(
+        const ColumnString::Chars & data,
+        const ColumnString::Offsets & offsets,
+        ColumnString::Chars & res_data,
+        ColumnString::Offsets & res_offsets)
+    {
+        const auto & encodings_freq = FrequencyHolder::getInstance().getEncodingsFrequency();
+
+        res_data.reserve(1024);
+        res_offsets.resize(offsets.size());
+
+        size_t prev_offset = 0;
+        size_t res_offset = 0;
+
+        for (size_t i = 0; i < offsets.size(); ++i)
+        {
+            const char * haystack = reinterpret_cast<const char *>(&data[prev_offset]);
+            String str = haystack;
+
+            String poss_ans;
+
+            std::unordered_map<UInt16, Float64> model;
+            calculateStats(str.data(), str.size(), readCodePoints, model);
+
+           Float64 max_result = log(zero_frequency) * (max_string_size);
+           for (auto& item : encodings_freq)
+            {
+                Float64 score = naiveBayes(item.map, model, max_result);
+                if (max_result < score)
+                {
+                    max_result = score;
+                    poss_ans = item.name;
+                }
+            }
+
+            size_t sep = poss_ans.find('_');
+            String ans_str;
+
+            if (detect_language)
+            {
+                ans_str = poss_ans.erase(0, sep + 1);
+            }
+            else
+            {
+                ans_str = poss_ans.erase(sep, poss_ans.size() - sep);
+            }
+
+            ans_str = poss_ans;
+
+            const auto res = ans_str.c_str();
+            size_t cur_offset = offsets[i];
+
+            size_t ans_size = strlen(res);
```

Comment:
strlen?? You have `ans_str.size()`

https://github.com/ClickHouse/ClickHouse/blob/515cc74530d11e1b2b18a63141b66a15b94748ba/src/Columns/ColumnString.h#L155-L165